--- ./github.com/fsnotify/fsnotify/fsnotify.go
+++ ./github.com/fsnotify/fsnotify/fsnotify.go
@@ -29,6 +29,7 @@
 	Remove
 	Rename
 	Chmod
+	CloseWrite
 )
 
 func (op Op) String() string {
@@ -50,6 +51,9 @@
 	if op&Chmod == Chmod {
 		buffer.WriteString("|CHMOD")
 	}
+	if op&CloseWrite == CloseWrite {
+		buffer.WriteString("|CLOSE_WRITE")
+	}
 	if buffer.Len() == 0 {
 		return ""
 	}
--- ./github.com/fsnotify/fsnotify/inotify.go
+++ ./github.com/fsnotify/fsnotify/inotify.go
@@ -96,7 +96,8 @@
 
 	const agnosticEvents = unix.IN_MOVED_TO | unix.IN_MOVED_FROM |
 		unix.IN_CREATE | unix.IN_ATTRIB | unix.IN_MODIFY |
-		unix.IN_MOVE_SELF | unix.IN_DELETE | unix.IN_DELETE_SELF
+		unix.IN_MOVE_SELF | unix.IN_DELETE | unix.IN_DELETE_SELF |
+		unix.IN_CLOSE_WRITE
 
 	var flags uint32 = agnosticEvents
 
@@ -333,5 +334,8 @@
 	if mask&unix.IN_ATTRIB == unix.IN_ATTRIB {
 		e.Op |= Chmod
 	}
+	if mask&unix.IN_CLOSE_WRITE == unix.IN_CLOSE_WRITE {
+		e.Op |= CloseWrite
+	}
 	return e
 }
