diff --git a/file/BUILD.bazel b/file/BUILD.bazel
index e01b831..a87edc3 100644
--- a/file/BUILD.bazel
+++ b/file/BUILD.bazel
@@ -15,6 +15,8 @@
 # Supporting infrastructure for implementing and testing PINS.
 #
 load("@com_github_grpc_grpc//bazel:cc_grpc_library.bzl", "cc_grpc_library")
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 
 package(
     default_visibility = ["//visibility:public"],
@@ -41,3 +43,23 @@ cc_grpc_library(
     grpc_only = True,
     deps = [":file_cc_proto"],
 )
+
+go_proto_library(
+    name = "file_go_proto",
+    compilers = ["@io_bazel_rules_go//proto:go_grpc"],
+    importpath = "github.com/openconfig/gnoi/file",
+    protos = [
+        ":file_proto",
+    ],
+)
+
+go_library(
+    name = "file",
+    embed = [":file_go_proto"],
+    importpath = "github.com/openconfig/gnoi/file",
+    deps = [
+        "//common:common",
+        "//types:types",
+    ],
+)
+
