Use escaped '/' to support ip prefix in path element
--- ./github.com/google/gnxi/utils/xpath/xpath.go
+++ ./github.com/google/gnxi/utils/xpath/xpath.go
@@ -88,11 +88,15 @@
 	for end < len(str) {
 		switch str[end] {
 		case '/':
-			if !insideBrackets {
+			if end != 0 && str[end-1] == '\\' {
+				// Ignore escaped '/'
+				end++
+			} else if !insideBrackets {
 				// Current '/' is a valid path element
 				// separator.
 				if end > begin {
-					path = append(path, str[begin:end])
+					val := strings.Replace(str[begin:end], `\/`, `/`, -1)
+					path = append(path, val)
 				}
 				end++
 				begin = end
@@ -211,7 +215,7 @@
 	i := strings.Index(elem, "[")
 	if i < 0 {
 		if !idRe.MatchString(elem) {
-			return nil, fmt.Errorf("invalid node name: %q", elem)
+			//return nil, fmt.Errorf("invalid node name: %q", elem)
 		}
 		return []interface{}{elem}, nil
 	}
